src/net/dns.rs

```diff
@@ -42,33 +82,107 @@ pub(crate) async fn prune_dns_cache(context: &Context) -> Result<()> {
     Ok(())
 }
 
-/// Looks up the hostname and updates DNS cache
-/// on success.
+/// Map from hostname to IP addresses.
+///
+/// NOTE: sync mutex is used, so it must not be held across `.await`
+/// to avoid deadlocks.
+static LOOKUP_HOST_CACHE: Lazy<std::sync::Mutex<HashMap<String, Vec<IpAddr>>>> =
+    Lazy::new(Default::default);
+
+async fn lookup_host_with_memory_cache(
+    context: &Context,
+    hostname: &str,
+    port: u16,
+) -> Result<Vec<IpAddr>> {
+    let stale_result = {
+        let mutex_guard = LOOKUP_HOST_CACHE.lock().unwrap();
+        mutex_guard.get(hostname).cloned()
+    };
+    if let Some(stale_result) = stale_result {
+        // Revalidate the cache in the background.
+        {
+            let context = context.clone();
+            let hostname = hostname.to_string();
+            tokio::spawn(async move {
+                match lookup_host((hostname.clone(), port))
+                    .await
+                    .context("DNS lookup failure")
+                {
+                    Ok(res) => {
+                        let res = res.into_iter().map(|addr| addr.ip()).collect();
+                        LOOKUP_HOST_CACHE.lock().unwrap().insert(hostname, res);
+                    }
+                    Err(err) => {
+                        warn!(
+                            context,
+                            "Failed to revalidate results for {hostname:?}: {err:#}."
+                        );
+                    }
+                }
+            });
+        }
+
+        info!(
+            context,
+            "Using memory-cached DNS resolution for {hostname}."
+        );
+        Ok(stale_result)
+    } else {
+        info!(
+            context,
+            "No memory-cached DNS resolution for {hostname} available, waiting for the resolver."
+        );
+        let res: Vec<IpAddr> = lookup_host((hostname, port))
```

Comment:
Probably may be factored out into a separate function to use also in the revalidation brach above, only the logs are different. But this is minor